components/sensor/bl0940.rst

```diff
@@ -64,8 +63,201 @@ Configuration variables:
+- **energy_reference** (*Optional*, float): The calibration parameter for cumulative energy readings. Defaults to ``6120.6267056536`` as calculated using schema defaults.
+
+.. _bl0940-calibration:
+
+Calibration
+-----------
+
+There are two fundamental calibration parameters which are dependent on the hardware design: ``voltage_reference`` and ``current_reference``.
+
+These can be determined by inspecting the actual hardware design or by using an accurate voltage and current meter with a simple resistive load
+
+Available calibration options:
+
+.. code-block:: yaml
+
+    # schematic option defaults
+    reference_voltage: 1.218  # Vref = 1.218
+    resistor_shunt: 1         # RL = 1 mΩ
+    resistor_one: 0.51        # R1 = 0.51 kΩ
+    resistor_two: 1950        # R2 = 5 x 390 kΩ -> 1950 kΩ
+
+Default values are based on `BL0940_APPNote_TSSOP14_V1.04_EN.pdf <https://www.belling.com.cn/media/file_object/bel_product/BL0940/guide/BL0940_APPNote_TSSOP14_V1.04_EN.pdf>`__.
+
+By default all the reference values will be calculated using the above values in the exact order shown below.
+
+.. code-block:: cpp
+
+    this->voltage_reference_ = 79931 / this->vref_ * (this->r_one_ * 1000) / (this->r_one_ + this->r_two_);
+    this->current_reference_ = 324004 * this->r_shunt_ / this->vref_;
+
+    // if voltage reference and current_reference have been configured
+    this->power_reference_ = this->voltage_reference_ * this->current_reference_ * 4046 / 324004 / 79931;
+    // if not, we use this formula based on provided formula in the specification
+    this->power_reference_ = 4046 * this->r_shunt_ * this->r_one_ * 1000 / this->vref_ / this->vref_ / (this->r_one_ + this->r_two_);
+
+    this->energy_reference_ = this->power_reference_ * 3600000 / (1638.4 * 256);
+
+If ``legacy_mode`` is enabled the reference values will be set according to initial implementation.
+
+.. code-block:: cpp
+
+    this->voltage_reference_ = 33000;
+    this->current_reference_ = 275000;
+    this->power_reference_   = 1430;
+    this->energy_reference_  = 3.6e6 / 297;
+
+Any of the reference values may be manually configured; user-specified values will take precedence over the calculated/predefined values.
+
+Theoretically, configuring ``voltage_reference`` and ``current_reference`` should be enough to calibrate the sensors, but, for completeness, ``power_reference`` and ``energy_reference`` are also configurable.
+
+.. code-block:: yaml
+
+    # custom references
+    voltage_reference: 17158.9201389365
+    current_reference: 266013.136288998
+    power_reference:   713.104696500825
+    energy_reference:  6120.6267056536
+
+Online Calibration
+------------------
+
+It is possible to adjust the sensor readings without need to update sensor configuration using the provided number platform.
+
+.. code-block:: yaml
+
+    # Example configuration entry
+    sensor:
+      - platform: bl0940
+        id: bl0940_id
+        uart_id: uart_bus
+        ...
+
+    number:
```

Comment:
Ideally each entity type goes on its own page in the relevant folder. eg: `/components/number/bl0940.rst`